network/src/noise_wrapper/handshake.rs

```diff
@@ -0,0 +1,356 @@
+        let (_, session) = self
+            .0
+            .finalize_connection(initiator_state, &server_response)
+            .map_err(|e| {
+                io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!("noise: wrong message received {}", e),
+                )
+            })?;
+
+        // finalize the connection
+        Ok(NoiseStream::new(socket, session))
+    }
+
+    pub async fn accept<TSocket>(
+        &self,
+        mut socket: TSocket,
+        anti_replay_timestamps: Option<Arc<RwLock<AntiReplayTimestamps>>>,
+        trusted_peers: Option<&Arc<RwLock<HashMap<PeerId, NetworkPeerInfo>>>>,
+    ) -> io::Result<NoiseStream<TSocket>>
+    where
+        TSocket: AsyncRead + AsyncWrite + Unpin,
+    {
+        // receive the initiation message
+        let mut client_init_message = [0u8; noise::handshake_init_msg_len(PAYLOAD_SIZE)];
+        socket.read_exact(&mut client_init_message).await?;
+
+        // parse it
+        let (their_public_key, handshake_state, payload) = self
+            .0
+            .parse_client_init_message(&[], &client_init_message)
+            .map_err(|e| {
+                io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!("noise: wrong message received {}", e),
+                )
+            })?;
+
+        // make sure the public key is a validator before continuing (if we're in the validator network)
+        if let Some(trusted_peers) = trusted_peers {
+            let found = trusted_peers
+                .read()
+                .unwrap()
+                .iter()
+                .any(|(_peer_id, public_keys)| public_keys.identity_public_key == their_public_key);
+            if !found {
+                // TODO: security logging (mimoo)
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!(
+                        "noise: client connecting to us with an unknown public key: {}",
+                        their_public_key
+                    ),
+                ));
+            }
+        }
+
+        // if on a mutually authenticated network
+        if let Some(anti_replay_timestamps) = &anti_replay_timestamps {
+            // check that the payload received as the client timestamp (in seconds)
+            if payload.len() != PAYLOAD_SIZE {
+                // TODO: security logging (mimoo)
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    "noise: client initiated connection without an 8-byte timestamp",
+                ));
+            }
+            let mut client_timestamp = [0u8; PAYLOAD_SIZE];
+            client_timestamp.copy_from_slice(&payload);
+            let client_timestamp = u64::from_le_bytes(client_timestamp);
+
+            // check the timestamp is not a replay
+            let mut anti_replay_timestamps = anti_replay_timestamps.write().unwrap();
```

Comment:
This is returning a `Result`, would it be appropriate to replace the `unwrap()` by a `?`, here?

Reply:
yes! Thanks for catching this : ) 

the only reason this would fail is because the lock has become poisoned. I don't see any reason this would happen since we don't run this in threads, but if it ever can happen we have no reason to panic and we should just log and continue to use our current connections.